crashtracker/src/collector/crash_handler.rs

```diff
@@ -41,37 +274,49 @@ enum ReceiverType {
 // `Box::from_raw` to recreate the box, then dropping it.
 static ALTSTACK_INIT: AtomicBool = AtomicBool::new(false);
 static OLD_HANDLERS: AtomicPtr<OldHandlers> = AtomicPtr::new(ptr::null_mut());
-static RECEIVER: AtomicPtr<ReceiverType> = AtomicPtr::new(ptr::null_mut());
 static METADATA: AtomicPtr<(CrashtrackerMetadata, String)> = AtomicPtr::new(ptr::null_mut());
 static CONFIG: AtomicPtr<(CrashtrackerConfiguration, String)> = AtomicPtr::new(ptr::null_mut());
-
-fn make_receiver(config: &CrashtrackerReceiverConfig) -> anyhow::Result<std::process::Child> {
-    // TODO: currently create the file in write mode.  Would append make more sense?
-    let stderr = if let Some(filename) = &config.stderr_filename {
-        File::create(filename)?.into()
-    } else {
-        Stdio::null()
-    };
-
-    let stdout = if let Some(filename) = &config.stdout_filename {
-        File::create(filename)?.into()
-    } else {
-        Stdio::null()
-    };
-
-    let receiver = Command::new(&config.path_to_receiver_binary)
-        .args(&config.args)
-        .envs(config.env.clone())
-        .stdin(Stdio::piped())
-        .stderr(stderr)
-        .stdout(stdout)
-        .spawn()
-        .context(format!(
-            "Unable to start process: {}",
-            config.path_to_receiver_binary
-        ))?;
-
-    Ok(receiver)
+static RECEIVER_CONFIG: AtomicPtr<CrashtrackerReceiverConfig> = AtomicPtr::new(ptr::null_mut());
+
+fn make_receiver(config: &CrashtrackerReceiverConfig) -> anyhow::Result<Receiver> {
+    let stderr = open_file_or_quiet(config.stderr_filename.as_deref())?;
+    let stdout = open_file_or_quiet(config.stdout_filename.as_deref())?;
+
+    // Create anonymous Unix domain socket pair for communication
+    let (uds_parent, uds_child) = socket::socketpair(
+        socket::AddressFamily::Unix,
+        socket::SockType::Stream,
+        None,
+        socket::SockFlag::empty(),
+    )
+    .context("Failed to create Unix domain socket pair")
+    .map(|(a, b)| (a.into_raw_fd(), b.into_raw_fd()))?;
+
+    // We need to spawn a process without calling atfork handlers, since this is happening inside
+    // of a signal handler.  Moreover, preference is given to multiplatform-uniform solutions.
+    // Although `vfork()` is deprecated, the alternatives have limitations
+    // * `fork()` calls atfork handlers
+    // * There is no guarantee that `posix_spawn()` will not call `fork()` internally
+    // * `clone()`/`clone3()` are Linux-specific
+    // Accordingly, use `vfork()` for now
+    match unsafe { vfork() } {
```

Comment:
so here the trade-off is we sacrifice macos compatibility
again thanks for the good comments

Reply:
Well, originally I chose `vfork()` purely because I understood that it was deprecated in macos, but still available.  It seems that it's dropped in macos from `nix`.  I need to do some investigation here.

If `vfork` really isn't available in macos, then I'll probably rewrite this implementation to spawn using the `clone()` syscall directly (I think it amounts to the same thing in this precise case, but `vfork()` generates warnings in Rust because the compiler does not have stable `#[ffi_returns_twice]`).

Reply:
But to be clear, the safety guarantees we currently strive for on Linux will not necessarily extend to macos.

Reply:
(Should we care a lot for macOS, at least as a first pass? SSI doesn't support macOS, so we could choose to enable crashtracker only on Linux for now)

Reply:
I don't think it needs to be a top-level concern, but a lot of folks do their initial testing on macos and it's nice for them to be able to see that their code-level changes worked in the happy case.

I don't anticipate it being a big deal.  Either I can pull up `vfork()` and everything works out-of-the-box, or I can't and I just call `fork()` on macos and `clone()` on linux.  :shrug:

Reply:
(ofc this would create a situation where the macos implementation is less correct, and possibly even responsible for spurious failures if we looked at it too hard, but I consider that an acceptable compromise for the developer ergonomics)

((these messages are written speculatively because I do not develop on macos and while I am sensitive to those who expect a common systems interface, I do not think we or anyone can provide such a guarantee within such a fundamentally weird use-case, so I do think there is some amount of risk/error in even trying this on macos in the first place))

Reply:
(((this message brought to you by the Lots of Isolated Silly Parenthesis society)))

Reply:
((((I approve of this message. Also I approve of LISP))))

Reply:
As someone who develops on osx and then does a final round of testing on linux, I'd prefer OSX to continue working for that workflow :).

Reply:
I'm just sayin': supporting for our own development is not the same as production-level bulletproof for customers :)

Reply:
(Or at least, not necessarily, depending on the effort required)

Reply:
> As someone who develops on osx and then does a final round of testing on linux, I'd prefer OSX to continue working for that workflow :).
Yup, sounds like we're on the same page.

> I'm just sayin': supporting for our own development is not the same as production-level bulletproof for customers :)
And it sounds like we're also on the same page.

Not sure how many folks can be on a single page, but given that I'm a 2-bit developer whose opinion doesn't matter 1 bit, that leaves at least 4095 bytes for the two of you, so I think we're fine.

Reply:
If this doesn't run on macOS, it blocks integrators (which pretty much all use macOS) from testing the crashtracker as a native extension in their languages without adding a lot of custom tooling. It would make development a lot more painful.

Reply:
Builds on mac, tests pass on mac.